src/middleware/devtools.ts

```diff
@@ -43,8 +43,8 @@ type StoreDevtools<S> = S extends {
   setState: (...a: infer A) => infer Sr
 }
   ? {
-      setState(
-        ...a: [...a: TakeTwo<A>, actionType?: string | { type: unknown }]
+      setState<AT extends string | { type: unknown; [key: string]: unknown }>(
```

Comment:
This is still incorrect, we don't need or want an index signature at all... Also let's rename `A` to `Sa` to avoid name clashing and let the public generic be `A`...

```diff
 type StoreDevtools<S> = S extends {
-  setState: (...a: infer A) => infer Sr
+  setState: (...a: infer Sa) => infer Sr
 }
   ? {
-      setState(
+      setState<A extends string | { type: unknown }>(
-        ...a: [...a: TakeTwo<A>, actionType?: string | { type: unknown }]
+        ...a: [...a: TakeTwo<Sa>, action?: A]
       ): Sr
     }
   : never
```

Reply:
Good point, TY.

Reply:
When we remove index signature we still won't be able to add action's payload: `Argument of type '{ type: string; count: number; }' is not assignable to parameter of type 'string | { type: unknown; } | undefined'.`

Reply:
That's not true, I tried making the change locally on your branch works fine. You're probably not making the exact change I showed above. Make the change and push we'll see if the CI complains.

Reply:
Ok, my bad. I didn't mention that the error occurs if devtools are used with immer:
```ts
const createBearSlice: StateCreator<
    { bears: { count: number } },
    [['zustand/devtools', never], ['zustand/immer', never]],
    [],
    { count: number }
  > = (set) => ({
    count: 0,
    setBears: (count: number) =>
      set(
        (state) => {
          state.bears.count = count
        },
        false,
        { type: 'bears/setBears', count } // <-- TS2345 
      ),
  })
```
I added some changes also to `immer.ts`. Could you take a look at them?

Reply:
It's recommended to use `immer` middleware before `devtools`, so that would fix your problem. You can change the test to use `immer` before `devtools`. I'll also open a PR to add this recommendation in the typescript doc.

As for the change you made in immer types, it's not correct, nor is it needed, so please revert it.

If we want to support using `devtools` before `immer` than the solution is much more complex. I'll see if I can open a PR for it later, but for now we don't need to support using `devtools` before `immer`

Also please don't make changes that you don't understand, it's okay if you don't fully understand the types here, it's better to ask for a solution than to guess a solution.

Reply:
> It's recommended to use immer middleware before devtools, so that would fix the problem. You can change the test to use immer before devtools. I'll also open a PR to add this recommendation in the typescript doc.

Ok, didn't know.

> As for the change you made in immer types, it's not correct, nor is it needed, so please revert it.
>
>If we want to support using devtools before immer than the solution is much more complex. I'll see if I can open a PR for it later, but for now we don't need to support using devtools before immer
>
>Also please don't make changes that you don't understand, it's okay if you don't fully understand the types here, it's better to ask for a solution than to guess a solution.

No offense, but it is also good to explain what mistake a person made, not just to point out that they did something wrong. Cause for me now it infers the type well while omitting the first two if the third is given. (Maybe I understood something wrong but definitely not guessing the "right" solution). It looks like you know the whole code very well so please explain why we should use immer before devtools and what is wrong with my commit?

Reply:
Only mistakes can be corrected like I did here https://github.com/pmndrs/zustand/pull/1183#issuecomment-1213132016. But the change you made is entirely incorrect, there's nothing to correct there. So first you will have to tell why you think it's correct, and then I can tell what is wrong in your understanding. But unfortunately, I'm not available for such OSS interactions (hence I didn't ask you why you think it's correct), especially when I can see the person did a guesswork (ie make a change to the code they don't understand). And it's okay if you don't think it was guesswork.

For now I can give you just one false positive case... With your change, `immer`'s `setState` will accept an action even when `devtools` is not present... The following code should not compile, but with your change it'll.

```ts
import create from 'zustand'
import { immer } from 'zustand/middleware'

interface BearState {
  bears: number
  increase: (by: number) => void
}

const useBearStore = create<BearState>()(immer((set) => ({
      bears: 0,
      increase: (by) => set((state) => ({ bears: state.bears + by }), false, "increase"),
    }
))
```

If you want to understand the types then see #710. Using immer before devtools is recommended precisely so that we don't have to do anything extra in immer types to support cases like this.

No offense taken or meant.

Reply:
Ok, I expressed myself wrong, it would be good to give an example when such a solution fails, which you did, so thank you. Now I see why it's a bad solution. Thanks for pointing out #710 I will dig into it. To be honest, these typings which you did is not the easiest to understand without fully digging into them (which can be time-consuming), but it's definitely a commendable effort of work (+1). It would be good to add that issue and PR in the documentation for development.

Reply:
> it would be good to give an example when such a solution fails

Yes I kinda agree, hence I made the edit...

> Maybe I should have pointed this out before itself, my apologies

> To be honest, these typings which you did is not the easiest to understand without fully digging into them (which can be time-consuming)

Yes, and hence users or contributors are not required to understand them. Only if one wishes to make a significant types contribution then it's required to understand them.

> it's definitely a commendable effort of work (+1)

Thanks!

> It would be good to add that issue and PR in the documentation for development.

It's actually already there in the typescript doc in the ["advanced usage" section](https://github.com/pmndrs/zustand/blob/main/docs/typescript.md#authoring-middlewares-and-advanced-usage).

Reply:
> It's actually already there in the typescript doc in the "advanced usage" section.

Woah, my bad, didn't noticed that 😞

Reply:
No problems!